docs/proposals/lifecycle_events.md

```diff
@@ -0,0 +1,128 @@
+# Support execution of the commands based on lifecycle events
+
+## Abstract
+Devfile 2.0.0 allows binding of commands to specific lifecycle events: https://github.com/devfile/kubernetes-api/issues/32. These commands are meant to be executed when those specific lifecycle events are triggered.
+
+The lifecycle events that have been implemented in Devfile 2.0.0 are generic. Their effect/meaning might be slightly different in the context a workspace (Che) or a single project (odo). However, odo should aim to provide a similar user experience within the scope of a single project.
+
+## Motivation
+Devfiles support commands that can be triggered based on development lifecycle events. With the implementation of lifecycle events, stack creators will be able to leverage all the capabilities of devfiles when building the correct experience/features for their stacks. `odo` should support/execute these commands at appropriate times within `odo` development flow.
+
+## User Stories
+Each lifecycle event would be suited to individual user stories. We propose that at least one issue is made for each of them for tracking and implementation purposes.
+
+- Add support for `postStart` commands: [container initialization support](https://github.com/openshift/odo/issues/2936)
+- Add support for `preStop` commands: 
+
+## Design overview
+Some of the events might not be useful for `odo` to adopt. We should only support the ones that have a clear use-case, and then add more as other use-cases emerge.
+
+For odo, the most important events are the `postStart`, and the `preStop` because they have clear, reasonable use case within odo’s flow. There is no clear use case within odo for `preStart` and `postStop` today.
+
+- `preStart` - Do not support in odo
```

Comment:
We should support this odo. This can be easily implemented as an initContainer.

Reply:
@kadel Do you have a clear use case in mind for preStart, that can't be achieved by postStart? 

We did consider init containers - convert devfile command into a command string and pass it as the command for the init container. But, weren't sure it would make sense if the command was going to use the same build/run container. So left it out as a future piece of work.

Reply:
> @kadel Do you have a clear use case in mind for preStart, that can't be achieved by postStart?

Currently, I don't. 
But this is how it was defined in devfile. We should implement it. Supporting only half the events in odo is going to just confuse everyone who will try to create devfiles.
If there are no technical reasons not to implement it we should implement it.

Reply:
It's just not clear to me how we can handle cleanly in odo.  If we try to support this and someone specify a command that uses the build/run container for the `preStart`, how do you even handle that? This command is meant to be called before those container get started. In order to execute those command, we need to start the container.

Reply:
`preStart` needs to be executed before the workspace pod (in odo case the component) is started. This is exactly what `initContainer` does. Che is coming to implement this in the same way. Check discussion in https://github.com/devfile/kubernetes-api/issues/32

Reply:
I've tried to re-read : devfile/kubernetes-api#32 and found something that I had missed earlier:

Implementation of preStart commands may need execution of a workspace pod init container. That means that the component associated to the preStart command should not be part of the workspace pod. The same applies to postStop bindings.

There is a clear requirement that container used for preStart and postStart commands will not overlap with containers of the main workspace pod. This makes is a bit clearer and perhaps we can implement preStart as init container or take an approach similar to postStop and run them as normal container commands before we run any postStart commands. 

If there is no overlap in containers, there should not really be any difference in the outcome and it might be simpler to implement without init containers as we have all that we need for that already.

Reply:
> It's just not clear to me how we can handle cleanly in odo. If we try to support this and someone specify a command that uses the build/run container for the `preStart`, how do you even handle that? This command is meant to be called before those container get started. In order to execute those command, we need to start the container.

@elsony Please see my comment above. This is something that I had missed too, but the spec clearly states that there should not be any overlap of containers between main workspace containers (build, run, test etc.) and preStart, postStop.

Reply:
I think we should go with preStart => initContainers and only run them when the component is bieng created.

Reply:
> when the component is bieng created.

This part might be tricky. I'm not sure if it will be possible. The initContainer will be executed everytime Deployment is triggered. Which might be right, or not :-( depending on what the command is doing.

Reply:
I was thinking that we could add init containers to the deployment.yaml only for the first push. and consecutive push, we can generate the deployment without init containers. But these can be looked at in more details during implementation phase as each will be implemented under separate issue.

For now we can update the proposal to include:

`preStart` - execute commands via init containers
`postStop` - execute commands via standalone pod or job and delete it once complete


@elsony @kadel would this make sense.